Scripts/UI/Base/UserInterface.gd

```diff
@@ -0,0 +1,103 @@
+extends Control
+class_name UserInterface
+## UserInterface Class: [br]
+## A Base UI class to contain common UI features within the game such as:[br]
+## - Controller Ui Support
+## - viewport Text Reizing (currently requires text size override to be on.).[br]
+## - Translation Switching (Unimplemented).
+
+@export_group("Focus Start", "start")
+
+enum FOCUS_START_ENUM 
+{
+	## Activates when ready function is called.
+	ready,
+	## Activates when changed to visible.
+	visibility_changed,
+	## Activates when either ready or visible is called
+	both
+}
+
+@export var start_on: bool = true
+@export var start_focus: Control
+@export var start_MODE: FOCUS_START_ENUM
+@export_group("Text Resize", "resize")
+
+## Turns Text Resizing on or off.
+@export var resize_on:bool = true
+
+enum RESIZE_MODE_ENUM 
+{
+	## the text will be resized using both the width and the height property.
+	both = 0,
+	## the text will be resized using th width property.
+	width = 2,
+	## the height will be resized using the height property.
+	height = 1
+}
+
+## Setting for how the UserInterface handles font resizing.
+@export var resize_MODE: RESIZE_MODE_ENUM = RESIZE_MODE_ENUM.both
+
+## original viewport base size.
+@onready var base_size: Vector2 = Vector2(
+	ProjectSettings.get_setting("display/window/size/viewport_width"), 
+	ProjectSettings.get_setting("display/window/size/viewport_height")
+)
+
+## Dictionary to hold nodes with font size overrides
+var font_nodes: Dictionary = {}  
+
+
+
+func _ready() -> void:
+	# Store font size and node
+	for node: Node in UtilityFunctions.get_all_Children(self):
+		if node is Control:
+			if node.has_theme_font_size_override("font_size"):
+				font_nodes[node] = node.get("theme_override_font_sizes/font_size")
+			elif node.has_theme_font_size("font_size"):
+				font_nodes[node] = node.get_theme_font_size("font_size")
+	
+	get_tree().get_root().size_changed.connect(font_resize)
+	font_resize()
+	if start_MODE == FOCUS_START_ENUM.ready or start_MODE == FOCUS_START_ENUM.both:
+		focus_grab()
+
+## Resizes all of the child nodes fonts acording to the viewport size.
```

Comment:
Function documentation should be inline w/function declaration.

Reply:
i would argue that this is a bad idea for the same reason that you should not make a class documentation inline with the class name, but i will fix it.

Reply:
Otherwise it won't show up in the autogenerated documentation. I agree it's a bit scuffed, but just how it is atm. Might make an issue in the godot repo.